src/mlpack/methods/cf/normalization/user_mean_normalization.hpp

```diff
@@ -0,0 +1,157 @@
+ * @endcode
+ */
+class UserMeanNormalization
+{
+ public:
+  // Empty constructor.
+  UserMeanNormalization() { }
+
+  /**
+   * Normalize the data by subtracting user mean from each of existing ratings.
+   *
+   * @param data Input dataset in the form of coordinate list.
+   */
+  void Normalize(arma::mat& data)
+  {
+    const size_t userNum = arma::max(data.row(0)) + 1;
+    userMean = arma::vec(userNum, arma::fill::zeros);
+    // Number of ratings for each user.
+    arma::Row<size_t> ratingNum(userNum, arma::fill::zeros);
+
+    // Sum ratings for each user.
+    data.each_col([&](arma::vec& datapoint) {
+      const size_t user = (size_t) datapoint(0);
+      const double rating = datapoint(2);
+      userMean(user) += rating;
+      ratingNum(user) += 1;
+    });
+
+    // Calculate user mean and subtract user mean from ratings.
+    // Set user mean to 0 if the user has no rating.
+    for (size_t i = 0; i < userNum; i++)
+    {
+      if (ratingNum(i) != 0)
+        userMean(i) /= ratingNum(i);
+    }
+
+    data.each_col([&](arma::vec& datapoint) {
+      const size_t user = (size_t) datapoint(0);
+      datapoint(2) -= userMean(user);
+      // The algorithm omits rating of zero. If normalized rating equals zero,
+      // it is set to the smallest positive double value.
+      if (datapoint(2) == 0)
+        datapoint(2) = std::numeric_limits<double>::min();
+    });
+  }
+
+  /**
+   * Normalize the data by subtracting user mean from each of existing rating.
+   *
+   * @param cleanedData Sparse matrix data.
+   */
+  void Normalize(arma::sp_mat& cleanedData)
+  {
+    userMean = arma::vec(arma::mean(cleanedData, 0));
+
+    arma::sp_mat::iterator it = cleanedData.begin();
+    arma::sp_mat::iterator it_end = cleanedData.end();
+    for (; it != it_end; it++)
+    {
+      *it = *it - userMean(it.col());
+      // The algorithm omits rating of zero. If normalized rating equals zero,
+      // it is set to the smallest positive double value.
+      if (*it == 0)
+        *it = std::numeric_limits<double>::min();
+    }
+  }
+
+  /**
+   * Denormalize computed rating by adding user mean.
+   *
+   * @param user User ID.
+   * @param item Item ID.
+   * @param rating Computed rating before denormalization.
+   */
+  double Denormalize(const size_t user,
+                     const size_t /* item */,
+                     const double rating) const
+  {
+    return rating + userMean(user);
+  }
+
+  /**
+   * Denormalize computed rating by adding user mean.
+   *
+   * @param combinations User/Item combinations.
+   * @param predictions Predicted ratings for each user/item combination.
+   */
+  void Denormalize(const arma::Mat<size_t>& combinations,
+                   arma::vec& predictions) const
+  {
+    for (size_t i = 0; i < predictions.n_elem; i++)
+    {
+      const size_t user = combinations(0, i);
+      predictions(i) += userMean(user);
+    }
+  }
+
+  /**
+   * Return user mean.
+   */
+  arma::vec UserMean() const
```

Comment:
Do you think we could name this ``mean`` as well? it might be easier if we use the same name for all methods, or do you think this might be confusing? Also what do you think about using:

```c++
const arma::vec& Mean() const { return userMean; }
```

Reply:
Yeah it's better to name this `Mean()` and return const reference to the object. I didn't think this through. Thanks!